File: raffia-core/src/main/java/com/codepoetics/raffia/paths/segments/StructUpdater.java
package com.codepoetics.raffia.paths.segments;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.baskets.Visitor;

import java.math.BigDecimal;

abstract class StructUpdater implements Visitor<Basket> {
  @Override
  public Basket visitString(String value) {
    return Basket.ofString(value);
  }

  @Override
  public Basket visitBoolean(boolean value) {
    return Basket.ofBoolean(value);
  }

  @Override
  public Basket visitNumber(BigDecimal value) {
    return Basket.ofNumber(value);
  }

  @Override
  public Basket visitNull() {
    return Basket.ofNull();
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/paths/segments/StructUpdater.kt
package com.codepoetics.raffia.paths.segments

import com.codepoetics.raffia.baskets.ArrayContents
import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.baskets.PropertySet
import com.codepoetics.raffia.java.api.Updater

internal abstract class StructUpdater : Updater {
    override fun update(basket: Basket): Basket {
        if (basket.isArray()) {
            return updateArray(basket.asArray())
        }

        if (basket.isObject()) {
            return updateObject(basket.asObject())
        }

        return basket
    }

    protected abstract fun updateArray(contents: ArrayContents): Basket

    protected abstract fun updateObject(properties: PropertySet): Basket
}
